lib/maths/CBoostedTreeImpl.cc

```diff
@@ -1286,12 +1293,14 @@ bool CBoostedTreeImpl::selectNextHyperparameters(const TMeanVarAccumulator& loss
 
     bopt.add(parameters, meanLoss, lossVariance);
     if (3 * m_CurrentRound < m_NumberRounds) {
-        std::generate_n(parameters.data(), parameters.size(), [&]() {
-            return CSampling::uniformSample(m_Rng, 0.0, 1.0);
-        });
-
+        std::move(m_HyperparameterSamples[m_CurrentRound].begin(),
+                  m_HyperparameterSamples[m_CurrentRound].end(), parameters.data());
```

Comment:
I'd change the entry condition for this block to `if (m_CurrentRound < m_HyperparameterSamples.size())` otherwise we can easily get undefined behaviour if we change `initializeHyperparameterSamples` and miss here.

Reply:
Also I'd just use a copy here since the moved types are doubles, i.e.
```
std::copy(m_HyperparameterSamples[m_CurrentRound].begin(),
                  m_HyperparameterSamples[m_CurrentRound].end(), parameters.data());
```

Reply:
Also just realised we probably want to make:
```
TVector parameters{this->numberHyperparametersToTune()};
```
at the start of this function
```
TVector parameters{m_TunableHyperparameters.size()};
```
for the same reason.

Reply:
:+1:  Great comments! I made the change.